pkg/util/slowset.go

```diff
@@ -81,18 +81,25 @@ func (s *SlowSet) TimeRemaining(key string) time.Duration {
 	return 0
 }
 
+func (s *SlowSet) removeAllExpired() {
+	s.Lock()
+	defer s.Unlock()
+	for key, t := range s.workSet {
+		if time.Since(t) > s.retentionTime {
+			delete(s.workSet, key)
+		}
+	}
+}
+
 func (s *SlowSet) Run(stopCh <-chan struct{}) {
+	ticker := time.NewTicker(s.resyncPeriod)
+	defer ticker.Stop()
 	for {
 		select {
 		case <-stopCh:
 			return
-		default:
-			time.Sleep(s.resyncPeriod)
-			for key, t := range s.workSet {
-				if time.Since(t) > s.retentionTime {
-					s.Remove(key)
-				}
-			}
+		case <-ticker.C:
+			s.removeAllExpired()
```

Comment:
How about:

```
func (s *SlowSet) removeAllExpired() {
	s.Lock()
	defer s.Unlock()

	for key, t := range s.workSet {
		if time.Since(t) > s.retentionTime {
			delete(s.workSet, key)
		}
	}
}

func (s *SlowSet) Run(stopCh <-chan struct{}) {
	for {
		select {
		case <-stopCh:
			return
		default:
			time.Sleep(s.resyncPeriod)
			s.removeAllExpired()
		}
	}
}
```

I do not like the idea of introducing yet another ticker in `Run` function when we are already in a loop.

Reply:
Using ticker with the loop is the way shown in the [official example](https://pkg.go.dev/time#example-NewTicker). I made this change to exit `Run` immediately when stopCh is triggered, which is commonly expected by caller. I don't see any major downside (maybe 1 or 2 extra allocations for the ticker and the extra channel).

I can revert to `time.Sleep` if you still insist. Please confirm.

Reply:
It is probably fine eitherway. Lets merge this.